alica_engine/src/engine/scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,107 @@
+#include "engine/scheduler/Scheduler.h"
+#include "engine/AlicaEngine.h"
+
+#include <algorithm>
+
+namespace alica
+{
+
+namespace scheduler
+{
+
+Scheduler::Scheduler(int numberOfThreads, const alica::AlicaEngine* ae) : _ae(ae), _running(true)
+{
+    for (int i = 0; i < numberOfThreads; i++) {
+        _workers.push_back(new std::thread(&Scheduler::workerFunction, this));
+    }
+}
+
+Scheduler::~Scheduler()
+{
+    _running = false;
+    _jobQueue.clear();
+
+    for (auto worker : _workers) {
+        _workerCV.notify_all();
+        worker->join();
+        delete worker;
+    }
+}
+
+void Scheduler::schedule(std::shared_ptr<Job> job)
+{
+    if (job->scheduledTime.inNanoseconds() == 0) {
+        job->scheduledTime = _ae->getAlicaClock().now();
+    }
+
+    if (job->inExecution || _jobQueue.isScheduled(job)) {
+        return;
+    }
+
+    // schedule missing prerequisites
+    for (auto prerequisite : job->prerequisites) {
+        if (auto prerequisiteSharedPtr = prerequisite.lock())  {
+            schedule(prerequisiteSharedPtr);
+        }
+    }
+
+    _jobQueue.insert(job);
+    _workerCV.notify_one();
+}
+
+void Scheduler::workerFunction()
+{
+    while(_running.load()) {
+        std::shared_ptr<Job> job;
+        bool executeJob = true;
+
+        {
+            std::unique_lock<std::mutex>lock(_workerMtx);
+            // wait when queue is empty. Do no wait when queue has jobs or the schedulers destructor is called.
+            _workerCV.wait(lock, [this]{return !_jobQueue.isEmpty() || !_running.load();});
+
+            if (_jobQueue.isEmpty() || !_running.load()) {
+                continue;
+            }
+
+            job = _jobQueue.popNext();
+
+            if (job->cancelled) {
+                /*
+                 * If job is cancelled:
+                 *  - Dont execute job.
+                 *  - Dont schedule job.
+                 */
+                continue;
+            }
+
+            job->inExecution = true;
+
+            for (std::weak_ptr<Job> prerequisite : job->prerequisites) {
```

Comment:
This loop is literally the only reason for the prerequisites to exist.
You could get rid of it by one of the following:
- Limiting this to a single worker thread. From what I can see, there is very little work being done, so that strategyu may be viable.
- Give each job a 64-bit id that is incremented. Have an atomic for each worker thread noting which id they are working on (if any). Use a single 64-bit to denote the largest job id that has to be finished before a job can start. Check only that id.

Unrelated:
this check should be part of the lambda predicated in the `wait()` above.

Reply:
Hey Hendrik, 
good to hear from you :D 

The general idea is to also schedule the behaviours init, run, and terminate methods with this scheduler. So 1 thread won't be enough. Further can one 64-bit ID denote that a job needs to wait for several other jobs to be finished?

This is, for example, the case when you switch from one state to another (A->B): The init of a behaviour in B should only be executed, when all terminate-functions of all plans and behaviours in A are done.

Reply:
:D thanks!
Yes this can be done if the ids are ordered and you wait for the maximum one.

Reply:
Okay, but there are no dependencies among the prerequisite jobs. So for example:
prerequisites: 1,2,3
actual job: 4
So now with four threads all get started, but job 4 will start once 3 is done. Without waiting for 1 and 2, since they have no relation to 3, but 3 is the one with the biggest ID. 

So I am still not convinced, can you see my point?

Reply:
No, it would wait for all jobs with an ID `<= 3`. In that system you would occasionally wait for too many jobs (the ones that go scheduled by the operation beforehand), but in practice that is probably negligible, and maybe even a good thing.
(Repeating jobs should be in another datastructure anyway).
To summarize, the proposal is:
- every job has an id
- every job has the maximum id of the jobs it is waiting on (0 if none).
- the scheduler keeps track of the largest consecutive id that is finished 
- optionally, job that you know noone is waiting for, can have a special invalid id.
- since there are thread-many jobs in execution, you will need something like:
```
std::atomic<int64_t> consecutive_done_counter;
std::array<std::atomic<int64_t>, kThreadCount> current_execution;
``` 
when a job is done you need to check if any id between `consecutive_done_counter` and `current_execution[i]` is mentioned in the other slots of `current_execution`. If not, you can safely set `consecutive_done_counter`. Be mindful about racecondition with getting/setting those atomics though.

Reply:
Nope, i don't think so. Whenever you are done you just check whether or not there is a job in execution between `consecutive_done_counter` and your current job id. If not, you increase `consecutive_done_counter` to the job id that just finished.

Reply:
Okay, back to my example then, but following your approach:

1. Jobs 1,2,3, and 4 are inserted into the queue. Job 4 waits on job 3. Job 1,2, and 3 don't wait for any other job. `consecutive_done_counter == 0`.
2. 3 threads take out the jobs 1-3 and start execute them, therefore current_execution includes ids 1,2, and 3.
3. Job 3 finishes first and since cannot update the `consecutive_done_counter`, since 1 and 2 are still in  `current_execution`, so `consecutive_done_counter` remains to be `0`.
4. Jobs 1 and 2 finish in order and update `consecutive_done_counter` to be 1 and 2 consecutively. So `consecutive_done_counter == 2`
5. *Huston we have problem:* Job 4 wont be executed since `consecutive_done_counter != 3` and no one will ever update it to 3.

Therefore, my suggestion is that in Step 3 we add Job 3 in `execution_done` and in Step 4 Job 2 will update `consecutive_done_counter` to the biggest ID that can be found in `execution_done` without "leaving a gap in the sequence in the order of IDs".

Reply:
Yes you are right, but an array `execution_done` won't fix it, since you could play the same game with multiple tasks that happen to be executed by the same thread.
So it would either need to be a vector or whenever a job finishes, it sets `consecutive_done_counter` to the maximum possible value, which would be `min(jobs_in_execution)-1`.  This I think would work nicely, as long as you don't have out-of-order execution. With OOE, you'll probably need the vector version of `execution_done`

Reply:
Right, that is why I said in my second to last post that it must be `std::array<std::atomic<int64_t>, kThreadCount-1> execution_done`. @abhi1287 I would like to make this change an extra PR (cc: @Karasuma1412) But I think it is definitely an improvement, since it will make the whole thing more lightweight -> no shared_ptr/weak_ptr, no explicit prerequisites

Reply:
My comment was that an array is not enough. You would need a vector.
Or the simplification I mentioned above.

I would advise against merging the current system as is, might as well fix the issues and then proceed. (But hey, your call) ;)

Reply:
The suggested approach has a problem that a single blocking app function call could keep everything waiting since the consecutive_ids are only incremented sequentially.  It might become a case that to decide whether a job can be executed, we will have to look at the consecutive_done_counter  and then look at this vector to determine whther something can be scheduled and all prerequisites are satisfied. 

example: 
Job id - {Prerequisite}
       1  - {}
       2  - {}
       3  - {}
       4 - {3}
       5 - {3}
       6 - {4}
       7 - {5}
       8 - {5}
consecutive_done_counter = 3   ->  4 & 5 will be scheduled.
Now if 4 happens to be a blocking call,  7 and 8 will be blocked from scheduling and so does any other children of 5.

Reply:
Yes good point. This would only work if all callback are reasonable. I don't know if you can make that assumption.

Reply:
That being said, these prerequisite vectors are dauntingly inefficient for a JobQueue.

Given what you want to do (call custom, mutable init/terminate functions), can you not hook into the RunningPlan state management more?
The activation function has a fixed order, and clean up also happens in a deterministic way.